cmake/modules/antlr4cpp.cmake

```diff
@@ -2,13 +2,52 @@ message(STATUS "antlr4cpp - using bundled version")
 
 include(ExternalProject)
 
+# check if uuid-dev is installed on the system, since this is a dependency of antlr4cpp
+find_package(PkgConfig)
+pkg_check_modules(UUID QUIET uuid)
+
+# if uuid-dev was not found on system we install it locally
+if (NOT UUID_FOUND)
+    message(STATUS "UUID not found - using bundled version")
+
+    set(LIBUUID_INSTALL_DIR "${CMAKE_BINARY_DIR}/uuid/install")
+    set(LIBUUID_PKGCONFIG_DIR ${LIBUUID_INSTALL_DIR}/lib/pkgconfig:$ENV{PKG_CONFIG_PATH})
+    set(LIBUUID_LIBRARY_DIR ${LIBUUID_INSTALL_DIR}/lib:$ENV{LIBRARY_PATH})
+    set(UUID_CONFIG_PARAMS "--prefix=${LIBUUID_INSTALL_DIR}")
+    ExternalProject_Add(
+        uuid_bundled
+        PREFIX uuid
+        URL ${PROJECT_SOURCE_DIR}/libs/libuuid-1.0.3.zip
+        URL_HASH MD5=9fd1e87682d24d6ca22941e0af339c8a
+        BUILD_IN_SOURCE TRUE
+        INSTALL_DIR "install"
+        CONFIGURE_COMMAND "./configure" ${UUID_CONFIG_PARAMS}
+        BUILD_COMMAND ${MAKE_EXE}
+    )
+    
+else()
+    set(LIBUUID_INSTALL_DIR "")
+    set(LIBUUID_PKGCONFIG_DIR "")
+    set(LIBUUID_LIBRARY_DIR "")
+    message(STATUS "UUID found - using system version")
+    # add a dummy target so the dependency in antlr4cpp_bundled is fulfilled if uuid-dev was found on the system
+    add_custom_target(uuid_bundled)
+endif ()
+
 ExternalProject_ADD(
         antlr4cpp_bundled
-        PREFIX      antlr4cppPrefix
-        URL         ${PROJECT_SOURCE_DIR}/libs/antlr4-cpp-runtime-4.9.3-source.zip
-        URL_HASH    MD5=eafa4fef583e12e963062882773461be
+        PREFIX           antlr4cppPrefix
+        URL              ${PROJECT_SOURCE_DIR}/libs/antlr4-cpp-runtime-4.9.3-source.zip
+        URL_HASH         MD5=eafa4fef583e12e963062882773461be
+        # pass PKG_CONFIG_PATH as a environment variable to cmake so find_package() in antlr4cpp's CMakeLists.txt can find uuid-dev if using the bundled version
+        CMAKE_COMMAND    ${CMAKE_COMMAND} -E env PKG_CONFIG_PATH=${LIBUUID_PKGCONFIG_DIR} ${CMAKE_COMMAND}
         BUILD_BYPRODUCTS ${CMAKE_BINARY_DIR}/antlr4cppPrefix/install/lib/libantlr4-runtime.a
-        CMAKE_ARGS  -DCMAKE_INSTALL_PREFIX=${CMAKE_BINARY_DIR}/antlr4cppPrefix/install -DCMAKE_CXX_FLAGS=-w
+        CMAKE_ARGS       -DCMAKE_INSTALL_PREFIX=${CMAKE_BINARY_DIR}/antlr4cppPrefix/install -DCMAKE_CXX_FLAGS=-w
+        # pass LIBRARY_PATH as a environment variable to make so the linker finds uuid-dev if using the bundled version
+        BUILD_COMMAND    ${CMAKE_COMMAND} -E env LIBRARY_PATH=${LIBUUID_LIBRARY_DIR} ${CMAKE_MAKE_PROGRAM}
+        LOG_BUILD        ON
+        LOG_INSTALL      ON
+        DEPENDS          uuid_bundled
```

Comment:
Check if the target can be substituted by a string that either holds the target name or is empty

Reply:
I need the `DEPENDS` here to ensure that `uuid` is present before the `ExternalProject` for `antlr4cpp` is configured. Without the `DEPENDS` I had problems when I use for example `make -j8`. However, `DEPENDS` expects a target, so I need to create the dummy target above. With a empty string it did not work.